File: src/test/java/no/novari/flyt/configuration/validation/ConfigurationValidatorFactoryTest.java
package no.novari.flyt.configuration.validation;

import no.novari.flyt.configuration.kafka.InstanceMetadataRequestProducerService;
import no.novari.flyt.configuration.kafka.IntegrationRequestProducerService;
import no.novari.flyt.configuration.kafka.MetadataRequestProducerService;
import no.novari.flyt.configuration.model.integration.Integration;
import no.novari.flyt.configuration.model.metadata.InstanceMetadataContent;
import no.novari.flyt.configuration.model.metadata.IntegrationMetadata;
import org.hibernate.validator.HibernateValidatorContext;
import org.hibernate.validator.HibernateValidatorFactory;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;

import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import java.util.Collections;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Answers.RETURNS_DEEP_STUBS;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.mockito.ArgumentMatchers.any;

class ConfigurationValidatorFactoryTest {

    @Mock
    private ValidatorFactory validatorFactory;
    @Mock
    private IntegrationRequestProducerService integrationRequestProducerService;
    @Mock
    private MetadataRequestProducerService metadataRequestProducerService;
    @Mock
    private InstanceMetadataRequestProducerService instanceMetadataRequestProducerService;

    private ConfigurationValidatorFactory configurationValidatorFactory;

    @BeforeEach
    void setUp() {
        integrationRequestProducerService = mock(IntegrationRequestProducerService.class, RETURNS_DEEP_STUBS);
        metadataRequestProducerService = mock(MetadataRequestProducerService.class, RETURNS_DEEP_STUBS);
        instanceMetadataRequestProducerService = mock(InstanceMetadataRequestProducerService.class, RETURNS_DEEP_STUBS);
        validatorFactory = mock(ValidatorFactory.class, RETURNS_DEEP_STUBS);
        configurationValidatorFactory = new ConfigurationValidatorFactory(validatorFactory, integrationRequestProducerService, metadataRequestProducerService, instanceMetadataRequestProducerService);

        HibernateValidatorFactory hibernateValidatorFactory = mock(HibernateValidatorFactory.class);
        HibernateValidatorContext hibernateValidatorContext = mock(HibernateValidatorContext.class);

        when(validatorFactory.unwrap(HibernateValidatorFactory.class)).thenReturn(hibernateValidatorFactory);
        when(hibernateValidatorFactory.usingContext()).thenReturn(hibernateValidatorContext);
        when(hibernateValidatorContext.constraintValidatorPayload(any())).thenReturn(hibernateValidatorContext);
        when(hibernateValidatorContext.getValidator()).thenReturn(mock(Validator.class));
    }

    @Test
    void getValidatorTest() {
        Integration mockIntegration = Integration.builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("id")
                .destination("destination")
                .state(Integration.State.ACTIVE)
                .activeConfigurationId("configId")
                .build();

        IntegrationMetadata mockMetadata = IntegrationMetadata.builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("id")
                .sourceApplicationIntegrationUri("uri")
                .integrationDisplayName("name")
                .version(1L)
                .build();

        InstanceMetadataContent mockInstanceMetadata = InstanceMetadataContent.builder()
                .categories(Collections.emptyList())
                .instanceObjectCollectionMetadata(Collections.emptyList())
                .instanceValueMetadata(Collections.emptyList())
                .build();

        when(integrationRequestProducerService.get(any())).thenReturn(Optional.of(mockIntegration));
        when(metadataRequestProducerService.get(any())).thenReturn(Optional.of(mockMetadata));
        when(instanceMetadataRequestProducerService.get(any())).thenReturn(Optional.of(mockInstanceMetadata));

        Validator result = configurationValidatorFactory.getValidator(1L, 1L);

        assertNotNull(result);
    }

    @Test
    void getValidatorIntegrationNotFoundTest() {
        when(integrationRequestProducerService.get(anyLong())).thenReturn(Optional.empty());

        assertThrows(CouldNotFindIntegrationException.class, () -> configurationValidatorFactory.getValidator(1L, 1L));
    }

    @Test
    void getValidatorMetadataNotFoundTest() {
        Integration mockIntegration = Integration.builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("id")
                .destination("destination")
                .state(Integration.State.ACTIVE)
                .activeConfigurationId("configId")
                .build();

        when(integrationRequestProducerService.get(any())).thenReturn(Optional.of(mockIntegration));
        when(metadataRequestProducerService.get(any())).thenReturn(Optional.empty());

        assertThrows(CouldNotFindMetadataException.class, () -> configurationValidatorFactory.getValidator(1L, 1L));
    }

    @Test
    void getValidatorInstanceMetadataNotFoundTest() {
        Integration mockIntegration = Integration.builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("id")
                .destination("destination")
                .state(Integration.State.ACTIVE)
                .activeConfigurationId("configId")
                .build();

        IntegrationMetadata mockMetadata = IntegrationMetadata.builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("id")
                .sourceApplicationIntegrationUri("uri")
                .integrationDisplayName("name")
                .version(1L)
                .build();

        when(integrationRequestProducerService.get(any())).thenReturn(Optional.of(mockIntegration));
        when(metadataRequestProducerService.get(any())).thenReturn(Optional.of(mockMetadata));
        when(instanceMetadataRequestProducerService.get(any())).thenReturn(Optional.empty());

        assertThrows(CouldNotFindInstanceMetadataException.class, () -> configurationValidatorFactory.getValidator(1L, 1L));
    }

}

File: src/test/kotlin/no/novari/flyt/configuration/validation/ConfigurationValidatorFactoryTest.kt
package no.novari.flyt.configuration.validation

import jakarta.validation.Validator
import jakarta.validation.ValidatorFactory
import no.novari.flyt.configuration.kafka.InstanceMetadataRequestProducerService
import no.novari.flyt.configuration.kafka.IntegrationRequestProducerService
import no.novari.flyt.configuration.kafka.MetadataRequestProducerService
import no.novari.flyt.configuration.model.integration.Integration
import no.novari.flyt.configuration.model.metadata.InstanceMetadataContent
import no.novari.flyt.configuration.model.metadata.IntegrationMetadata
import org.hibernate.validator.HibernateValidatorContext
import org.hibernate.validator.HibernateValidatorFactory
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Answers
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

class ConfigurationValidatorFactoryTest {
    private lateinit var validatorFactory: ValidatorFactory
    private lateinit var integrationRequestProducerService: IntegrationRequestProducerService
    private lateinit var metadataRequestProducerService: MetadataRequestProducerService
    private lateinit var instanceMetadataRequestProducerService: InstanceMetadataRequestProducerService

    private lateinit var configurationValidatorFactory: ConfigurationValidatorFactory

    @BeforeEach
    fun setUp() {
        integrationRequestProducerService =
            mock(defaultAnswer = Answers.RETURNS_DEEP_STUBS)
        metadataRequestProducerService =
            mock(defaultAnswer = Answers.RETURNS_DEEP_STUBS)
        instanceMetadataRequestProducerService =
            mock(defaultAnswer = Answers.RETURNS_DEEP_STUBS)
        validatorFactory = mock(defaultAnswer = Answers.RETURNS_DEEP_STUBS)
        configurationValidatorFactory =
            ConfigurationValidatorFactory(
                validatorFactory,
                integrationRequestProducerService,
                metadataRequestProducerService,
                instanceMetadataRequestProducerService,
            )

        val hibernateValidatorFactory: HibernateValidatorFactory = mock()
        val hibernateValidatorContext: HibernateValidatorContext = mock()

        whenever(validatorFactory.unwrap(HibernateValidatorFactory::class.java)).thenReturn(hibernateValidatorFactory)
        whenever(hibernateValidatorFactory.usingContext()).thenReturn(hibernateValidatorContext)
        whenever(hibernateValidatorContext.constraintValidatorPayload(any<ConfigurationValidationContext>()))
            .thenReturn(hibernateValidatorContext)
        whenever(hibernateValidatorContext.validator).thenReturn(mock<Validator>())
    }

    @Test
    fun getValidatorTest() {
        val mockIntegration =
            Integration
                .builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("id")
                .destination("destination")
                .state(Integration.State.ACTIVE)
                .activeConfigurationId("configId")
                .build()

        val mockMetadata =
            IntegrationMetadata
                .builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("id")
                .sourceApplicationIntegrationUri("uri")
                .integrationDisplayName("name")
                .version(1L)
                .build()

        val mockInstanceMetadata =
            InstanceMetadataContent
                .builder()
                .categories(emptyList())
                .instanceObjectCollectionMetadata(emptyList())
                .instanceValueMetadata(emptyList())
                .build()

        whenever(integrationRequestProducerService.get(any())).thenReturn(mockIntegration)
        whenever(metadataRequestProducerService.get(any())).thenReturn(mockMetadata)
        whenever(instanceMetadataRequestProducerService.get(any())).thenReturn(mockInstanceMetadata)

        val result = configurationValidatorFactory.getValidator(1L, 1L)

        assertNotNull(result)
    }

    @Test
    fun getValidatorIntegrationNotFoundTest() {
        whenever(integrationRequestProducerService.get(any())).thenReturn(null)

        assertThrows(CouldNotFindIntegrationException::class.java) {
            configurationValidatorFactory.getValidator(1L, 1L)
        }
    }

    @Test
    fun getValidatorMetadataNotFoundTest() {
        val mockIntegration =
            Integration
                .builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("id")
                .destination("destination")
                .state(Integration.State.ACTIVE)
                .activeConfigurationId("configId")
                .build()

        whenever(integrationRequestProducerService.get(any())).thenReturn(mockIntegration)
        whenever(metadataRequestProducerService.get(any())).thenReturn(null)

        assertThrows(CouldNotFindMetadataException::class.java) {
            configurationValidatorFactory.getValidator(1L, 1L)
        }
    }

    @Test
    fun getValidatorInstanceMetadataNotFoundTest() {
        val mockIntegration =
            Integration
                .builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("id")
                .destination("destination")
                .state(Integration.State.ACTIVE)
                .activeConfigurationId("configId")
                .build()

        val mockMetadata =
            IntegrationMetadata
                .builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("id")
                .sourceApplicationIntegrationUri("uri")
                .integrationDisplayName("name")
                .version(1L)
                .build()

        whenever(integrationRequestProducerService.get(any())).thenReturn(mockIntegration)
        whenever(metadataRequestProducerService.get(any())).thenReturn(mockMetadata)
        whenever(instanceMetadataRequestProducerService.get(any())).thenReturn(null)

        assertThrows(CouldNotFindInstanceMetadataException::class.java) {
            configurationValidatorFactory.getValidator(1L, 1L)
        }
    }
}
